docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,573 @@
+resource scaling ([#10782](https://issues.k8s.io/10782)).
+
+### Implementation Plan
+
+#### 1.7 Alpha Release
+
+We're targeting an alpha release in Kubernetes 1.7 that includes the following
+basic functionality:
+
+*   Support in the kubelet for attaching debug containers to a running pod
+*   A `kubectl debug` command to initiate a debug container
+*   `kubectl describe pod` will list status of debug containers running in a pod
+
+Functionality in the kubelet will be hidden behind an alpha feature flag and
+disabled by default. The following are explicitly out of scope for the 1.7 alpha
+release:
+
+*   *Copy Debug* will not be implemented
+*   `kubectl describe pod` will not report the running command, only the
+    container status
+*   Garbage collection of exitted debug containers cannot be tuned
+*   Specific integration with cluster admission controller and pod security
+    policy, though these may work by default
+
+*Running Debug* is implemented in the kubelet's generic runtime manager.
+Performing this operation with a legacy (non-CRI) runtime or a cluster without
+the alpha feature enabled will result in a not implemented error. Implementation
+will be broken into the following steps:
+
+##### Step 1: Container Type
+
+Debug containers exist within the kubelet as a `ContainerStatus` without a
+container spec. As with other `kubecontainers`, their state is persisted in the
+runtime via labels.
+
+To distinguish debug containers, the runtime uses a new
+`io.kubernetes.container.type` label. Existing containers will be started with a
+type of `REGULAR`, `INIT` or `SANDBOX`. When added in a subsequent step, debug
+containers will start with with the type `DEBUG`. The `type` label will populate
+a new field `Type` in `container.ContainerStatus`.
+
+##### Step 2: Creation and Handling of Debug Containers
+
+This step adds methods for creating debug containers, but doesn't yet modify the
+kubelet API. The kubelet will gain a `RunDebugContainer()` method which accepts
+a `v1.Container` and creates a debug container. Fields that don't make sense in
+the context of a debug container (e.g. probes, lifecycle, resources) will be
+excluded by a white list and ignored.
+
+The kubelet will treat `DEBUG` containers differently in the following ways:
+
+1.  `SyncPod()` ignores containers of type `DEBUG`, since there is no
+    configuration to sync.
+1.  `DEBUG` containers will be excluded from calculation of pod phase and
+    condition.
+1.  `DEBUG` containers will not be garbage collected while the pod exists.
+
+All containers will continue to be killed by `KillPod()`, which already operates
+on containers returned by the runtime and does not discriminate on type.
+
+##### Step 3: kubelet API changes
+
+The kubelet will gain a new streaming endpoint `/debug` similar to `/exec`,
+`/attach`, etc. The debug endpoint will create a new debug container and
+automatically attach to it.
```

Comment:
Is it allowed to create a non-interactive debug container?

Also, creating and starting a container may take a long time (e.g., pulling down the image). Wouldn't this affect the user experience?